### PR TITLE
WIP: Add Quantum Kernel Training Runtime Client

### LINKED ISSUE #228: Implement the QKA runtime

### PR DESCRIPTION
**Pull request by calebj15**
Add QKA runtime client to Qiskit.
Closes #228.

### CODE REVIEW

File: qiskit_machine_learning/utils/loss_functions/kernel_loss_functions.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        """
+        This method calls the ``evaluate`` method. This is a convenient method to compute loss.
+        """
+        return self.evaluate(parameter_values, quantum_kernel, data, labels)
+
+    def get_variational_callable(
+        self, quantum_kernel: QuantumKernel, data: Iterable, labels: Iterable
+    ) -> Callable[[Sequence[float]], float]:
+        """
+        Return a callable variational loss function given some inputs. The sole input to the
```
Comment (calebj15) [inline qiskit_machine_learning/utils/loss_functions/kernel_loss_functions.py:51]: reword this to reflect latest

File: qiskit_machine_learning/utils/loss_functions/kernel_loss_functions.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+from ...kernels import QuantumKernel
+
+
+class KernelLoss(ABC):
+    """
+    Abstract base class for computing the loss of a kernel function.
+    Unlike many loss functions, which only take into account the labels and predictions
+    of a model, kernel loss functions may be a function of internal model parameters or
+    quantities that are generated during training. For this reason, extensions of this
+    class may find it necessary to introduce additional inputs.
```
Comment (calebj15) [inline qiskit_machine_learning/utils/loss_functions/kernel_loss_functions.py:32]: Remove last sentence

File: qiskit_machine_learning/kernels/algorithms/quantum_kernel_trainer.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+            raise ValueError(f"Unknown loss {loss}!")
+
+    @property
+    def optimizer(self) -> Optimizer:
+        """Returns an optimizer to be used in training."""
+        return self._optimizer
+
+    @optimizer.setter
+    def optimizer(self, optimizer: Optimizer) -> None:
+        """Sets the loss."""
```
Comment (calebj15) [inline qiskit_machine_learning/kernels/algorithms/quantum_kernel_trainer.py:142]: "sets the optimizer"

File: qiskit_machine_learning/kernels/algorithms/quantum_kernel_trainer.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+        self,
+        data: np.ndarray,
+        labels: np.ndarray,
+    ) -> QuantumKernelTrainerResult:
+        """
+        Train the QuantumKernel by minimizing loss over the kernel parameters. The input
+        quantum kernel will not be altered, and an optimized quantum kernel will be returned.
+
+        Args:
+            data (numpy.ndarray): ``NxD`` array of training data, where ``N`` is the
```
Comment (calebj15) [inline qiskit_machine_learning/kernels/algorithms/quantum_kernel_trainer.py:165]: (N, D) and (N, 1)

File: test/runtime/test_quantum_kernel_trainer_client.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+        qkt_program = QuantumKernelTrainerClient(
+            quantum_kernel=quantum_kernel,
+            backend=backend,
+            loss=loss_func,
+            optimizer=optimizer,
+            shots=self.shots,
+            provider=provider,
+        )
+        qkt_program.fit_kernel(self.sample_train, self.label_train)
```
Comment (calebj15) [inline test/runtime/test_quantum_kernel_trainer_client.py:89]: Finish this test

File: qiskit_machine_learning/kernels/algorithms/quantum_kernel_trainer.py [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+        # Return the QuantumKernel in optimized state
+        output_kernel.assign_user_parameters(result.optimal_parameters)
+        result.quantum_kernel = output_kernel
+
+        return result
+
+
+def _str_to_variational_callable(
+    loss_str: str,
+    quantum_kernel: "QuantumKernel" = None,
```
Comment (calebj15) [inline qiskit_machine_learning/kernels/algorithms/quantum_kernel_trainer.py:215]: Why is QK in quotes

File: qiskit_machine_learning/runtime/quantum_kernel_trainer_client.py [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+        """Set the quantum kernel."""
+        self._quantum_kernel = quantum_kernel
+
+    @property
+    def loss(self) -> Callable:
+        """Return the loss."""
+        return self._loss
+
+    @loss.setter
+    def loss(self, loss: Callable) -> None:
```
Comment (calebj15) [inline qiskit_machine_learning/runtime/quantum_kernel_trainer_client.py:157]: Make this match signature for QKT

File: qiskit_machine_learning/runtime/quantum_kernel_trainer_client.py [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+        """Sets the loss."""
+        if callable(loss):
+            self._loss = loss
+        else:
+            raise TypeError(
+                "The loss function must be an instance of ``KernelLoss`` or ``Callable``."
+            )
+
+    @property
+    def optimizer(self) -> Union[Optimizer, Dict[str, Any]]:
```
Comment (calebj15) [inline qiskit_machine_learning/runtime/quantum_kernel_trainer_client.py:167]: What is this return type?

File: qiskit_machine_learning/runtime/quantum_kernel_trainer_client.py [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+    def initial_layout(self, initial_layout: Optional[Iterable[int]]) -> None:
+        """Sets the initial layout."""
+        self._initial_layout = initial_layout
+
+    def fit_kernel(
+        self,
+        data: Iterable[float],
+        labels: Iterable[float],
+    ) -> QKTRuntimeResult:
+        """Calls the Quantum Kernel Training Runtime ('quantum-kernel-training') to train
```
Comment (calebj15) [inline qiskit_machine_learning/runtime/quantum_kernel_trainer_client.py:236]: quantum-kernel-trainer

File: qiskit_machine_learning/kernels/algorithms/quantum_kernel_trainer.py [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+        Train the QuantumKernel by minimizing loss over the kernel parameters. The input
+        quantum kernel will not be altered, and an optimized quantum kernel will be returned.
+
+        Args:
+            data (numpy.ndarray): ``(N, D)`` array of training data, where ``N`` is the
+                              number of samples and ``D`` is the feature dimension
+            labels (numpy.ndarray): ``(N, 1)`` array of +/-1 labels of the ``N`` training samples
+
+        Returns:
+            QuantumKernelTrainerResult: the results of kernel training
```
Comment (dlasecki) [inline qiskit_machine_learning/kernels/algorithms/quantum_kernel_trainer.py:170]:
```suggestion
            QuantumKernelTrainerResult: the results of kernel training.
```

File: qiskit_machine_learning/kernels/algorithms/quantum_kernel_trainer.py [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+
+        Raises:
+            ValueError: No trainable user parameters specified in quantum kernel
```
Comment (dlasecki) [inline qiskit_machine_learning/kernels/algorithms/quantum_kernel_trainer.py:173]:
```suggestion
            ValueError: No trainable user parameters specified in quantum kernel.
```

File: qiskit_machine_learning/kernels/quantum_kernel.py [hunk 12]
```diff
@@ -101,19 +101,19 @@ def feature_map(self, feature_map: QuantumCircuit) -> None:
 
     @property
     def unbound_feature_map(self) -> QuantumCircuit:
-        """Returns unbound feature map"""
+        """Return unbound feature map"""
```
Comment (dlasecki) [inline qiskit_machine_learning/kernels/quantum_kernel.py:104]:
```suggestion
        """Return unbound feature map."""
```

Comment (dlasecki) [reply qiskit_machine_learning/kernels/quantum_kernel.py:104]: Please check other places for missing dots.

File: qiskit_machine_learning/kernels/quantum_kernel.py [hunk 13]
```diff
@@ -172,30 +187,92 @@ def assign_user_parameters(
...
+        # in general. To resolve this issue, we handle each entry of the values dict one at a time.
+        for param, bind in values.items():
+            if isinstance(bind, ParameterExpression):
+                self._unbound_feature_map.assign_parameters({param: bind}, inplace=True)
+
+                # User params are all non-input params in the unbound feature map
+                # This list comprehension ensures that self._user_parameters is ordered
+                # in a way that is consistent with self.feature_map.parameters
+                self._user_parameters = [
+                    p for p in self._unbound_feature_map.parameters if (p not in input_params)
```
Comment (dlasecki) [inline qiskit_machine_learning/kernels/quantum_kernel.py:223]: The `in` check would work much faster if `input_params` was still a set.

File: qiskit_machine_learning/runtime/quantum_kernel_trainer_client.py [hunk 14]
```diff
@@ -0,0 +1,344 @@
...
+        if isinstance(loss, str):
+            self._loss = loss.lower()
+            if self._loss == "svc_loss":
+                pass
+            else:
+                raise ValueError(f"Unknown loss {loss}!")
+        elif callable(loss):
+            self._loss = loss  # type: ignore
+        else:
+            raise ValueError(f"Unknown loss { loss}!")
```
Comment (dlasecki) [inline qiskit_machine_learning/runtime/quantum_kernel_trainer_client.py:168]:
```suggestion
            raise ValueError(f"Unknown loss type { type(loss)}!")
```

File: qiskit_machine_learning/runtime/quantum_kernel_trainer_client.py [hunk 15]
```diff
@@ -0,0 +1,344 @@
...
+    @property
+    def optimizer(self) -> Optimizer:
+        """Return the optimizer."""
+        return self._optimizer
+
+    @optimizer.setter
+    def optimizer(self, optimizer: Optimizer) -> None:
+        """Set the optimizer."""
+        if not isinstance(optimizer, Optimizer):
+            raise TypeError("The optimizer must be an instance of qiskit.algorithms.optimizers")
```
Comment (dlasecki) [inline qiskit_machine_learning/runtime/quantum_kernel_trainer_client.py:179]: Might print provided type explicitly.

File: qiskit_machine_learning/utils/loss_functions/kernel_loss_functions.py [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+                    ``N = # samples, M = dimension of data``
+            labels: A length-N array containing the truth labels
+
+        Returns:
+            A loss value
+        """
+        raise NotImplementedError
+
+
+class SVCLoss(KernelLoss):
```
Comment (dlasecki) [inline qiskit_machine_learning/utils/loss_functions/kernel_loss_functions.py:80]: I would suggest putting classes to different files.

File: qiskit_machine_learning/utils/loss_functions/kernel_loss_functions.py [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+        dual_coefs = svc.dual_coef_[0]
+
+        # Get support vectors
+        support_vecs = svc.support_
+
+        # Prune kernel matrix of non-support-vector entries
+        kmatrix = kmatrix[support_vecs, :][:, support_vecs]
+
+        # Calculate loss
+        loss = np.sum(np.abs(dual_coefs)) - (0.5 * (dual_coefs.T @ kmatrix @ dual_coefs))
```
Comment (dlasecki) [inline qiskit_machine_learning/utils/loss_functions/kernel_loss_functions.py:141]: Might extract loss expression to a method `calc_loss` and remove the comment.

File: test/kernels/algorithms/test_qkernel_trainer.py [hunk 18]
```diff
@@ -0,0 +1,111 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
```
Comment (dlasecki) [inline test/kernels/algorithms/test_qkernel_trainer.py:3]: Remember to also add 2022 now. Also check other places.

File: test/runtime/fake_qkt_runtime.py [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+
+
+class FakeQKTRuntimeProvider(Provider):
+    """A fake runtime provider for unit tests."""
+
+    def has_service(self, service):
+        """Check if a service is available."""
+        if service == "runtime":
+            return True
+        return False
```
Comment (dlasecki) [inline test/runtime/fake_qkt_runtime.py:92]:
```suggestion
        return service == "runtime":
```

**Comment by calebj15** [comment]
This project is on hold until base runtimes are available.

When Estimator is available, we will use it to evaluate our kernel matrices.

Expect this to be picked up late Feb / early Mar

**Comment by adekusar-drl** [comment]
@calebj15 do you need this PR? May I close it as outdated?

**Comment by calebj15** [comment]
> @calebj15 do you need this PR? May I close it as outdated?

yes it can be closed, thanks
